File: go/src/1423.可获得的最大点数.go
/*
 * @lc app=leetcode.cn id=1423 lang=golang
 *
 * [1423] 可获得的最大点数
 */

// @lc code=start
func maxScore(cardPoints []int, k int) int {
	n := len(cardPoints)
	nums := make([]int, k + k)
	
	for i := 0; i < k; i++ {
		nums[i] = cardPoints[n - k + i]
	}
	for i := k; i < k + k; i++ {
		nums[i] = cardPoints[i - k]
	}

	cur, res := 0, 0
	for i := 0; i < k - 1; i++ {
		cur += nums[i]
	} 
	for i := k - 1; i < k + k; i++ {
		cur += nums[i]
		res = max(res, cur)
		cur -= nums[i - k + 1]
	}

	return res
}

func max(a ...int) int {
	res := a[0]
	for _, v := range a {
		if res < v {
			res = v
		}
	}
	return res
}
// @lc code=end

File: rust/src/1423.可获得的最大点数.rs
/*
 * @lc app=leetcode.cn id=1423 lang=rust
 *
 * [1423] 可获得的最大点数
 */

// @lc code=start
impl Solution {
    pub fn max_score(card_points: Vec<i32>, k: i32) -> i32 {
        let k = k as usize;
        let n = card_points.len();
        // 把首尾各k个元素 构造一个[n-k, n-1] + [0, k-1]
        let mut nums = vec![0; 2 * k as usize];
        for i in 0..k {
            nums[i] = card_points[n - k + i];
        }
        for i in k..k * 2 {
            nums[i] = card_points[i - k];
        }
        let mut res = 0;
        let mut cur = 0i64;
        for i in 0..k - 1 {
            cur += nums[i] as i64;
        }

        for i in (k - 1)..(k * 2) {
            cur += nums[i] as i64;
            res = res.max(cur as i32);
            cur -= nums[i + 1 - k] as i64;
        }

        res
    }
}
// @lc code=end
